### DISCUSSION TITLE
CreateRelationshipProperties and RelationshipProperties - usage

### DISCUSSION THREAD

**Discussion by guntarsTB** [themetalfleece/neogma#100] [Q&A]
Hello,

I'm trying to figure out proper usage of the CreateRelationshipProperties and RelationshipProperties which are type attributes on ModelRelatedNodesI.

```
/** used for defining the type of the RelatedNodesToAssociate interface, to be passed as the second generic to ModelFactory */
export interface ModelRelatedNodesI<
  /** the type of the related model */
  RelatedModel extends {
    createOne: NeogmaModelStaticsI<any>["createOne"];
  },
  /** the instance of the related model */
  RelatedInstance,
  /** properties for the relationship */
  CreateRelationshipProperties extends RelationshipPropertiesI | Object = Object,
  RelationshipProperties extends RelationshipPropertiesI | Object = Object> 
{
  /** interface of the data to create */
  CreateData: Parameters<RelatedModel["createOne"]>[0] & Partial<CreateRelationshipProperties>;
  /** interface of the properties of the relationship used in create functions */
  CreateRelationshipProperties: CreateRelationshipProperties;
  /** interface of the actual properties of the relationship */
  RelationshipProperties: RelationshipProperties;
  /** the instance of the related model */
  Instance: RelatedInstance;
}
```

So far all my attempts to figure it out have failed, and the best theoretical explanation is contradicting with the closest practical conclusions. By theory and intuition (including reading of the comments) it seems that CreateRelationshipProperties can define one set of attributes while RelationshipProperties can define another set of attributes. That eventually could allow different attributes for creation process than the actual attributes of the relationship. I can assume that creation of a relationship needs context attributes which will not be stored as properties, or the relationship itself can have calculated properties that are not passed as arguments during the creation. Either way, they could be different, and this could be designed to solve this problem.

On the other hand, the actual code doesn't seem to solve this problem. Furthermore, both sets of properties meet at the same place and start to intervene with each other at `RelationshipsI`.

Here, one type is used as a key, the other one is used as a property, and in real life it causes very strange behavior when playing with them.

```
/** the type for the Relationship configuration of a Model */
export type RelationshipsI<RelatedNodesToAssociateI extends AnyObject> = {
  /** the alias of the relationship definitions is the key */
  [alias in keyof RelatedNodesToAssociateI]: {
    /** the related model. It could be the object of the model, or "self" for this model */
    model: NeogmaModel<any, any, any, any> | 'self';
    /** the name of the relationship */
    name: CreateRelationshipParamsI['relationship']['name'];
    /** the direction of the relationship */
    direction: 'out' | 'in' | 'none';
    /** relationship properties */
    properties?: {
      /** the alias of the relationship property is the key */
      [relationshipPropertyAlias in keyof RelatedNodesToAssociateI[alias]['CreateRelationshipProperties']]: {
        /** the actual property to be used on the relationship */
        property: keyof RelatedNodesToAssociateI[alias]['RelationshipProperties'];
        /** validation for the property */
        schema: IValidationSchema;
      };
    };
  };
};
```

I am not pretending that I totally understand the whole logic, therefore I don't dare to claim it's an issue, but it seems like it could be.
Any explanation of the above would be highly appreciated, of if that's an issue, please take it for a record as such.

**Comment by themetalfleece** [comment]
Hey there!

You may find more info [here](https://themetalfleece.github.io/neogma/docs/Models/Defining-a-Model#using-modelfactory) if you click on the `Typescript` tab.
Basically, `CreateRelationshipProperties` is a type where each key is an alias of the properly to be created on the relationship, while `RelationshipProperties` is a type with the actual property names, as they will be stored in the database.

That being said, the comments could definitely be improved in the code.

Let me know if this answers your question!
